server.go

```diff
@@ -243,6 +245,42 @@ func (s *Server) register(sd *ServiceDesc, ss interface{}) {
 	s.m[sd.ServiceName] = srv
 }
 
+// Metadata returns the metadata for a given symbol name.
+// The name can be a service name or a method name in the form of
+// <package>.<service>[.<method>].
+func (s *Server) Metadata(name string) interface{} {
+	// Check if the name is a service name.
+	if srv, ok := s.m[name]; ok {
+		return srv.meta
+	}
+	// Check if the name is a method name.
+	pos := strings.LastIndex(name, ".")
+	if pos == -1 {
+		// Invalid method name.
+		return nil
+	}
+	if srv, ok := s.m[name[:pos]]; ok {
+		if _, ok := srv.md[name[pos+1:]]; ok {
+			return srv.meta
+		}
+		if _, ok := srv.sd[name[pos+1:]]; ok {
+			return srv.meta
+		}
+	}
+	return nil
+}
+
+// AllServiceNames returns names of all the registered services.
+func (s *Server) AllServiceNames() []string {
+	ret := make([]string, len(s.m))
```

Comment:
https://golang.org/ref/spec#Making_slices_maps_and_channels

make(T, n)       slice      slice of type T with length n and capacity n
make(T, n, m)    slice      slice of type T with length n and capacity m

You should never be using make([]type, arraylen, arraylen)